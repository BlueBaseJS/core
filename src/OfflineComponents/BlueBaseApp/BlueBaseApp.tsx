import { BlueBase, BlueBaseProgress, BootOptions } from '../../BlueBase';
import {
	BlueBaseAppError,
	BlueBaseAppErrorProps,
	BlueBaseAppLoading,
	BlueBaseAppLoadingProps,
} from '..';
import React, { useEffect, useState } from 'react';

import { BlueBaseProvider } from '../../Context';
import { IntlProvider } from '../../intl';
import { ThemeProvider } from '../../themes';
import { getComponent } from '../../getComponent';

const BlueBaseContent = getComponent('BlueBaseContent');

export interface BlueBaseAppProps extends Partial<BootOptions> {
	/** BlueBase Context */
	BB?: BlueBase;

	/** Component rendered when an error occurs */
	ErrorComponent: React.ComponentType<BlueBaseAppErrorProps>;

	/** Component rendered when an error occurs */
	LoadingComponent: React.ComponentType<BlueBaseAppLoadingProps>;

	/**
	 * If this prop is provided, this tree will be rendered instead of BlueBase's own view.
	 */
	children?: React.ReactNode;

	/**
	 * Used to locate this view in end-to-end tests.
	 */
	testID?: string;
}

/**
 * # ðŸš€ BlueBaseApp
 *
 * The main BlueBase app. This is the top level component in BlueBase. Takes care
 * of initialisation, and renders either children, or app with routing.
 *
 * ## Usage
 * ```jsx
 * <BlueBaseApp BB={BB} />
 * ```
 */
export const BlueBaseApp = (props: BlueBaseAppProps) => {
	const {
		ErrorComponent = BlueBaseAppError,
		LoadingComponent = BlueBaseAppLoading,
		children,
	} = props;

	const [bootCount, setBootCount] = useState(0);
	const [booting, setBooting] = useState(true);
	const [bootTrigger, setBootTrigger] = useState(true); // Setting to true to start boot
	const [progress, setProgress] = useState<BlueBaseProgress>({});

	const [BB] = useState(props.BB || new BlueBase());

	BB.reboot = async () => {
		setBootTrigger(true);
	};

	useEffect(() => {
		if (bootTrigger) {
			setBootTrigger(false);
		} else {
			return;
		}

		(async () => {
			await BB.boot({
				...props,
				onProgress: (p: BlueBaseProgress) => {
					setProgress(p);
					if (p.error) {
						setBooting(false);
					}
				},
			});
			setBootCount(bootCount + 1);
			setBooting(false);
		})();
	}, [bootTrigger]);

	if (booting) {
		return <LoadingComponent BB={BB} progress={progress} bootCount={bootCount} />;
	}

	if (progress.error) {
		return <ErrorComponent BB={BB} progress={progress} bootCount={bootCount} />;
	}

	return (
		<BlueBaseProvider key={`boot-${bootCount}`} value={BB}>
			<ThemeProvider>
				<IntlProvider>
					<BlueBaseContent>{children}</BlueBaseContent>
				</IntlProvider>
			</ThemeProvider>
		</BlueBaseProvider>
	);
};

const defaultProps: Partial<BlueBaseAppProps> = {
	ErrorComponent: BlueBaseAppError,
	LoadingComponent: BlueBaseAppLoading,
};

BlueBaseApp.defaultProps = defaultProps;
