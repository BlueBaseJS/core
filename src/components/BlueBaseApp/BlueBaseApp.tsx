import { BlueBase, BootOptions } from '../../BlueBase';
import React from 'react';
import { Text } from 'react-native';

export interface BlueBaseAppProps extends Partial<BootOptions> {
	BB?: BlueBase,
	children?: React.ReactNode,
}

interface BlueBaseAppState {

	booted: boolean;
	BB: BlueBase

	Component: React.ComponentType<any>
}

/**
 * ðŸš€ BlueBaseApp
 *
 * This is the Main App used to render at the top level.
 *
 * TODO: Add better docs
 */
// NOTE FOR DEVELOPERS:
// Don't use BlueBase context or any data saved in context (e.g. components)
// here. This is because the context may not be initialized yet.
export class BlueBaseApp extends React.Component<BlueBaseAppProps, BlueBaseAppState> {

	constructor(props: BlueBaseAppProps) {
		super(props);

		this.state = {
			BB: props.BB || new BlueBase(),
			Component: () => <Text>Loading</Text>,
			booted: false,
		};
	}

	async componentDidMount() {

		const BB = this.state.BB;

		try {
			const Component = await BB.boot(this.props);
			this.setState({
				Component,
				booted: BB.booted,
			});
		} catch (error) {
			// tslint:disable-next-line:no-console
			console.error(error);
		}
	}

	render() {

		const Component = this.state.Component;

		return <Component children={this.props.children} />;
	}
}