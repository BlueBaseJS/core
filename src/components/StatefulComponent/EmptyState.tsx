import { BlueRainType, withBlueRain } from '../../index';
import React from 'react';

const EmptyState = (props: { bluerain: BlueRainType }) => {

	const BR = props.bluerain;

	const ComponentState = BR.Components.get('ComponentState');

	return (
		<ComponentState
			title="Empty area"
			description="No data found."
		/>
	);
};

export default withBlueRain(EmptyState);
