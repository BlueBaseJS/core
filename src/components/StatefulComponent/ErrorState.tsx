import { BlueRainType,withBlueRain } from '../../index';
import Icon from '../Icon';
import React from 'react';

const ErrorIcon = () => <Icon title="Error" />;

const ErrorState = (props: { bluerain: BlueRainType }) => {

	const BR = props.bluerain;

	const ComponentState = BR.Components.get('ComponentState');

	return (
		<ComponentState
			title="Something broke!"
			description="An unknown error has occured. Please try again later."
			image={ErrorIcon}
		/>
	);
};

export default withBlueRain(ErrorState);
