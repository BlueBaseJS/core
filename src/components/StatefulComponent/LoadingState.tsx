import { BlueRain, withBlueRain } from '../../index';
import React from 'react';

const LoadingState = (props: { bluerain: BlueRain }) => {

	const BR = props.bluerain;

	const Indicator = () => <BR.Components.ActivityIndicator color="#eeeeee" size="large" />;

	return <BR.Components.ComponentState title="Loading, Please wait." image={Indicator} />;
};

export default withBlueRain(LoadingState);
