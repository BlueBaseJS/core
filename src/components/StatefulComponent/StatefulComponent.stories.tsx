import { BlueRain, withBlueRain } from '../../index';
import React from 'react';
import storiesOf from '../../../storybook/storiesOf';

storiesOf('StatefulComponent', module)
	.add('StatefulComponent children function prop', () => {

		const Story = withBlueRain((props: { bluerain: BlueRain }) => {
			const BR = props.bluerain;

			return (
				<BR.Components.StatefulComponent>
				{() =>{
					return <BR.Components.Text>Render prop pattern</BR.Components.Text>
				}}
				</BR.Components.StatefulComponent>
			);
		});

		return <Story />;
	})

	.add('StatefulComponent children prop', () => {

		const Story = withBlueRain((props: { bluerain: BlueRain }) => {
			const BR = props.bluerain;

			return (
				<BR.Components.StatefulComponent>
					<BR.Components.Text>This this a child!</BR.Components.Text>
				</BR.Components.StatefulComponent>
			);
		});

		return <Story />;
	})

	.add('StatefulComponent throw error', () => {

		const Story = withBlueRain((props: { bluerain: BlueRain }) => {
			const BR = props.bluerain;

			return (
				<BR.Components.StatefulComponent>
					{() => {
						throw new Error('Boom!');
					}}
				</BR.Components.StatefulComponent>
			);
		});

		return <Story />;
	})


	.add('Only loading Component', () => {

		const Story = withBlueRain((props: { bluerain: BlueRain }) => {
			const BR = props.bluerain;
			const LoadingState = BR.Components.get('LoadingState');

			return <LoadingState />;
		});

		return <Story />;
	})

    .add('Only EmptyState Component', () => {

	const Story = withBlueRain((props: { bluerain: BlueRain }) => {
		const BR = props.bluerain;
		const EmptyState = BR.Components.get('EmptyState');

		return <EmptyState  />;
	});

	return <Story />;
})

.add('Only ErrorState Component', () => {

	const Story = withBlueRain((props: { bluerain: BlueRain }) => {
		const BR = props.bluerain;
		const ErrorState = BR.Components.get('ErrorState');

		return <ErrorState  />;
	});

	return <Story />;
});


// storiesOf('ComponentState', module).add('Simple', () => <div>hello </div>);
