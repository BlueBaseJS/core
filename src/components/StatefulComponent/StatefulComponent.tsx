import {
	DataObserver,
	EmptyState,
	ErrorObserver,
	LoadingState,
	WaitObserver,
} from '../../getComponent';
import {
	DataObserverChildrenProps,
	StatefulComponentProps,
	WaitObserverChildrenProps,
} from '@bluebase/components';

import React from 'react';

// tslint:disable: jsdoc-format
/**
 * # üë®‚Äçüé® StatefulComponent
 *
 * This is a swiss army knife component. Intended to be used as a single source of UI
 * state management. It shows empty, loading, error or data states based on the given props.
 *
 * ## Usage
```jsx
<StatefulComponent data={data} loading={true} delay={200} timeout={10000}>
 <Text>Content</Text>
</StatefulComponent>
```
 */
export class StatefulComponent extends React.PureComponent<StatefulComponentProps> {
	public static defaultProps: Partial<StatefulComponentProps> = {
		emptyComponent: EmptyState,
		loadingComponent: LoadingState,
		timeout: 10000,
	};

	render() {
		const {
			component: Component,
			loadingComponent: LoadingComponent,
			emptyComponent: EmptyComponent,
			errorComponent,
			children,

			// DataObserver
			isLoading,
			isEmpty,
			loading,
			data,

			// WaitObserver
			delay,
			timeout,
			onRetry,
			onTimeout,

			// ErrorObserver
			error,
			checkError,

			...other
		} = this.props;

		const rest = { data, ...other };

		return (
			<ErrorObserver {...{ error, checkError, errorComponent, rest }}>
				<DataObserver {...{ isEmpty, isLoading, loading, data, rest }}>
					{(event: DataObserverChildrenProps) => {
						if (event.loading === true) {
							return React.createElement(WaitObserver, {
								children: (props: WaitObserverChildrenProps) => <LoadingComponent {...props} />,
								delay,
								onRetry,
								onTimeout,
								timeout,
							});
						}

						if (event.empty) {
							return <EmptyComponent />;
						}

						// Render 'component' prop
						if (Component) {
							return React.createElement(Component, rest);
						}

						// 'children' as a function, 'render prop' pattern
						if (typeof children === 'function') {
							return (children as any)(rest);
						}

						// children
						return children;
					}}
				</DataObserver>
			</ErrorObserver>
		);
	}
}
