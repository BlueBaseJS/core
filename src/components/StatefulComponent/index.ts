import StatefulComponent, { StatefulComponentProperties } from './StatefulComponent';

import EmptyState from './EmptyState';
import ErrorState from './ErrorState';
import LoadingState from './LoadingState';

export default StatefulComponent;
export { EmptyState, ErrorState, LoadingState };

export { StatefulComponentProperties };
