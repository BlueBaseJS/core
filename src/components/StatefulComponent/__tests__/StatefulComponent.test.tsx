import { BlueBaseApp } from '../../BlueBaseApp';
import React from 'react';
import { StatefulComponent } from '../StatefulComponent';
import { Text } from 'react-native';
import { mount } from 'enzyme';
import { waitForElement } from 'enzyme-async-helpers';
// import TestRenderer from 'react-test-renderer';

// jest.useFakeTimers();

describe('StatefulComponent', () => {

	test(`should show empty state as no data is provided`, async () => {
		const component = mount(
			<BlueBaseApp>
				<StatefulComponent />
			</BlueBaseApp>
		);

		await waitForElement(component as any, StatefulComponent);

		// expect(component).toMatchSnapshot();

		expect(component.find('StatefulComponent H6 Text').last().text()).toBe('Empty Area');
		expect(component.find('StatefulComponent Body2 Text').last().text()).toBe('Create new items to fill this area.');
	});

	test(`should show empty state as data is an empty array`, async () => {

		const component = mount(
			<BlueBaseApp>
				<StatefulComponent data={[]} />
			</BlueBaseApp>
		);

		await waitForElement(component as any, StatefulComponent);

		// expect(component).toMatchSnapshot();

		expect(component.find('StatefulComponent H6 Text').last().text()).toBe('Empty Area');
		expect(component.find('StatefulComponent Body2 Text').last().text()).toBe('Create new items to fill this area.');

	});

	test(`should show loading state if loading is set to true`, async () => {
		const component = mount(
			<BlueBaseApp>
				<StatefulComponent loading={true} delay={0} timeout={0} />
			</BlueBaseApp>
		);

		await waitForElement(component as any, StatefulComponent);

		// expect(component).toMatchSnapshot();
		expect(component.exists('LoadingState')).toBe(true);
	});

	it(`should show children nodes`, async () => {
		const component = mount(
			<BlueBaseApp>
				<StatefulComponent data={true}>
					<Text>Content</Text>
				</StatefulComponent>
			</BlueBaseApp>
		);

		await waitForElement(component as any, StatefulComponent);

		// expect(component).toMatchSnapshot();
		expect(component.find('StatefulComponent').find('Text').last().text()).toBe('Content');
	});

	it(`should show render prop children`, async () => {
		const component = mount(
			<BlueBaseApp>
				<StatefulComponent data={true}>
					{() => <Text>Render prop</Text>}
				</StatefulComponent>
			</BlueBaseApp>
		);

		await waitForElement(component as any, StatefulComponent);

		// expect(component).toMatchSnapshot();
		expect(component.find('StatefulComponent').find('Text').last().text()).toBe('Render prop');
	});

	it(`should show children from Component prop`, async () => {

		const Comp = () => <Text>Custom Component</Text>;

		const component = mount(
			<BlueBaseApp>
				<StatefulComponent data={true} component={Comp}>
					<Text>Content</Text>
				</StatefulComponent>
			</BlueBaseApp>
		);

		await waitForElement(component as any, StatefulComponent);

		// expect(component).toMatchSnapshot();
		expect(component.find('StatefulComponent').find('Text').last().text()).toBe('Custom Component');
	});

});
