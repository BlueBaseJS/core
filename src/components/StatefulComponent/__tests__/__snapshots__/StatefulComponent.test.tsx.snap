// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StatefulComponent Snapshot StatefulComponent component 1`] = `
<Text>
  Loading
</Text>
`;

exports[`StatefulComponent Snapshot StatefulComponent component after complete render 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
    }
  }
>
  <Text
    style={
      Array [
        Object {
          "color": "rgba(0, 0, 0, 0.87)",
          "fontFamily": "Roboto",
          "fontSize": 20,
          "fontWeight": "500",
        },
        Object {
          "marginBottom": 8,
          "textAlign": "center",
        },
      ]
    }
  >
    Empty area
  </Text>
  <Text
    style={
      Array [
        Object {
          "color": "rgba(0, 0, 0, 0.87)",
          "fontFamily": "Roboto",
          "fontSize": 16,
          "fontWeight": "400",
        },
        Object {
          "textAlign": "center",
        },
      ]
    }
  >
    No data found.
  </Text>
</View>
`;

exports[`StatefulComponent Snapshot StatefulComponent component after complete render with child 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
    }
  }
>
  <Text
    style={
      Array [
        Object {
          "color": "rgba(0, 0, 0, 0.87)",
          "fontFamily": "Roboto",
          "fontSize": 20,
          "fontWeight": "500",
        },
        Object {
          "marginBottom": 8,
          "textAlign": "center",
        },
      ]
    }
  >
    Empty area
  </Text>
  <Text
    style={
      Array [
        Object {
          "color": "rgba(0, 0, 0, 0.87)",
          "fontFamily": "Roboto",
          "fontSize": 16,
          "fontWeight": "400",
        },
        Object {
          "textAlign": "center",
        },
      ]
    }
  >
    No data found.
  </Text>
</View>
`;

exports[`StatefulComponent Snapshot StatefulComponent component after complete render with child, isEmpty: () => true 1`] = `null`;

exports[`StatefulComponent Snapshot StatefulComponent component after complete render with child, isEmpty: () => true, Component 1`] = `null`;

exports[`StatefulComponent Snapshot StatefulComponent component after complete render with child, loading: true 1`] = `null`;
